### PR TITLE
F/library

### PR DESCRIPTION
**Pull request by jinal-apnamart**

### CODE REVIEW

File: GeofencingModule/build.gradle [hunk 1]
```diff
@@ -1,58 +1,58 @@
...
+        sourceCompatibility = JavaVersion.VERSION_11
+        targetCompatibility = JavaVersion.VERSION_11
     }
     kotlinOptions {
-        jvmTarget = '11'
+        jvmTarget = "11"
     }
 }
-group = 'com.github.vilas-apnamart'
+group = "com.github.jinal-apnamart"
```
Comment (vilas-apnamart) [inline GeofencingModule/build.gradle]: change this back

File: GeofencingModule/build.gradle [hunk 2]
```diff
@@ -1,58 +1,58 @@
...
 
     defaultConfig {
-        minSdk 26
+        minSdk = 26
 
-        testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
+        testInstrumentationRunner = "androidx.test.runner.AndroidJUnitRunner"
         consumerProguardFiles "consumer-rules.pro"
         versionCode 1
         versionName "1.0.0"
```
Comment (vilas-apnamart) [inline GeofencingModule/build.gradle]: update this

File: GeofencingModule/src/main/java/com/apnamart/geofencingmodule/geofencing/library/GeofenceLibrary.kt [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+import com.apnamart.geofencingmodule.geofencing.event_handler.GeofenceEventHandler
+import com.apnamart.geofencingmodule.geofencing.provider.GeofenceDataProvider
+import com.apnamart.geofencingmodule.geofencing.work_manager.AddGeofenceWorker
+import com.apnamart.geofencingmodule.geofencing.work_manager.WorkManagerInitializer
+import com.apnamart.geofencingmodule.geofencing.work_manager.worker_utils.scheduleOneTimeWorkerWithOutData
+import com.apnamart.geofencingmodule.geofencing.work_manager.worker_utils.schedulePeriodicWorkerWithConstraints
+import java.lang.ref.WeakReference
+import java.util.concurrent.TimeUnit
+
+object GeofenceLibrary {
```
Comment (vilas-apnamart) [inline GeofencingModule/src/main/java/com/apnamart/geofencingmodule/geofencing/library/GeofenceLibrary.kt]: rename this

File: GeofencingModule/src/main/java/com/apnamart/geofencingmodule/geofencing/library/GeofenceLibrary.kt [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    internal fun getGeofenceDataProvider(): GeofenceDataProvider? {
+        return dataProvider
+    }
+
+    internal fun getGeofenceManager(): GeofenceManager? {
+        return geofenceManager
+    }
+
+    fun cancelGeofenceWorkers() {
+        workManager?.cancelAllWorkByTag(GeofenceConstants.GEOFENCE_SERVICE_WORKER_JOB)
```
Comment (vilas-apnamart) [inline GeofencingModule/src/main/java/com/apnamart/geofencingmodule/geofencing/library/GeofenceLibrary.kt]: remove geofences

File: GeofencingModule/src/main/java/com/apnamart/geofencingmodule/geofencing/permissions/LocationPermissionHelper.kt [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+     * Checks if the required location permissions are granted.
+     *
+     * @param context The context to check permissions against.
+     * @return A list of missing permissions, if any.
+     */
+    fun checkLocationPermissions(context: Context): Boolean {
+        val fineAndCoarseLocation = checkFineAndCoarseLocationPermission(context)
+        if (!fineAndCoarseLocation) {
+            return false
+        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q && !checkBackgroundLocationPermission(context)) {
```
Comment (vilas-apnamart) [inline GeofencingModule/src/main/java/com/apnamart/geofencingmodule/geofencing/permissions/LocationPermissionHelper.kt]: make if if

File: GeofencingModule/src/main/java/com/apnamart/geofencingmodule/geofencing/permissions/LocationPermissionHelper.kt [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        ))
+    }
+
+    @RequiresApi(Build.VERSION_CODES.Q)
+    fun checkBackgroundLocationPermission(context: Context): Boolean{
+        return checkPermission(context, Manifest.permission.ACCESS_BACKGROUND_LOCATION)
+    }
+
+    fun createLocation(provider: String, lat: Double, long: Double): Location {
+        val location = Location(provider)
```
Comment (vilas-apnamart) [inline GeofencingModule/src/main/java/com/apnamart/geofencingmodule/geofencing/permissions/LocationPermissionHelper.kt]: why is this here?

File: GeofencingModule/src/main/java/com/apnamart/geofencingmodule/geofencing/provider/GeofenceDataProvider.kt [hunk 7]
```diff
@@ -0,0 +1,23 @@
+package com.apnamart.geofencingmodule.geofencing.provider
+
+import com.apnamart.geofencingmodule.geofencing.data.GeofenceData
+
+
+interface GeofenceDataProvider {
+
+    /**
+     * Method to check if geofence should be added or not.
+     * The app will implement this to provide validation checks to determine if geofence should eb added or not.
```
Comment (vilas-apnamart) [inline GeofencingModule/src/main/java/com/apnamart/geofencingmodule/geofencing/provider/GeofenceDataProvider.kt]: typo in should be

File: GeofencingModule/src/main/java/com/apnamart/geofencingmodule/geofencing/work_manager/AddGeofenceWorker.kt [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+        val geofenceDataProvider = GeofenceLibrary.getGeofenceDataProvider()
+        val geofenceManager = GeofenceLibrary.getGeofenceManager()
+
+        if (geofenceDataProvider == null || geofenceManager == null) {
+            return Result.failure()
+        }
+
+        if (!geofenceDataProvider.shouldAddGeofence()) {
+            Log.e(GeofenceConstants.TAG, "geofence should not be added")
```
Comment (vilas-apnamart) [inline GeofencingModule/src/main/java/com/apnamart/geofencingmodule/geofencing/work_manager/AddGeofenceWorker.kt:30]: remove logs

File: GeofencingModule/src/main/java/com/apnamart/geofencingmodule/geofencing/work_manager/WorkManagerInitializer.kt [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+class WorkManagerInitializer(private val context: WeakReference<Context>) {
+
+    val workManager by lazy {
+        context.get()?.let {
+            try {
+                // Attempt to get the WorkManager instance
+                Log.e(TAG, "WorkManager instance already initialised")
+                WorkManager.getInstance(it)
+            } catch (e: IllegalStateException) {
+                Log.e(TAG, "WorkManager not initialized: ${e.message}")
```
Comment (vilas-apnamart) [inline GeofencingModule/src/main/java/com/apnamart/geofencingmodule/geofencing/work_manager/WorkManagerInitializer.kt:20]: remove logs

File: app/build.gradle [hunk 10]
```diff
@@ -1,48 +1,42 @@
...
-    id 'com.android.library'
-    id 'org.jetbrains.kotlin.android'
+    alias(libs.plugins.android.library)
+    alias(libs.plugins.jetbrains.kotlin.android)
 }
 
 android {
-    namespace 'com.apnamart.geofencing_module'
-    compileSdk 34
+    namespace = "com.apnamart.geofencingmodule"
```
Comment (vilas-apnamart) [inline app/build.gradle]: ??

File: jitpack.yml [hunk 11]
```diff
@@ -5,4 +5,4 @@ before_install:
   - ./scripts/prepareJitpackEnvironment.sh
 install:
   - FILE="-Dfile=geofencing-module.aar"
-  - mvn install:install-file $FILE -DgroupId=com.github.vilas-apnamart -DartifactId=geofencing-module -Dversion=1.0.0 -Dpackaging=aar -DgeneratePom=true
\ No newline at end of file
+  - mvn install:install-file $FILE -DgroupId=com.github.jinal-apnamart -DartifactId=geofencing-module -Dversion=1.0.0 -Dpackaging=aar -DgeneratePom=true
```
Comment (vilas-apnamart) [inline jitpack.yml]: undo this

File: pom.xml [hunk 12]
```diff
@@ -2,7 +2,7 @@
 <project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns="http://maven.apache.org/POM/4.0.0"
     xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
     <modelVersion>4.0.0</modelVersion>
-    <groupId>com.github.vilas-apnamart</groupId>
+    <groupId>com.github.jinal-apnamart</groupId>
```
Comment (vilas-apnamart) [inline pom.xml]: undo this

File: settings.gradle [hunk 13]
```diff
@@ -12,6 +12,6 @@ pluginManagement {
     }
 }
 
-rootProject.name = "geofencing-module"
-include ':app'
-include ':GeofencingModule'
+rootProject.name = "GeofencingModule"
```
Comment (vilas-apnamart) [inline settings.gradle]: undo this

**Review by vilas-apnamart — CHANGES_REQUESTED** [review]

File: GeofencingModule/build.gradle (hunk 1 above)
Comment (jinal-apnamart) [reply GeofencingModule/build.gradle]: done

File: GeofencingModule/build.gradle (hunk 2 above)
Comment (jinal-apnamart) [reply GeofencingModule/build.gradle]: done

File: GeofencingModule/src/main/java/com/apnamart/geofencingmodule/geofencing/library/GeofenceLibrary.kt (hunk 3 above)
Comment (jinal-apnamart) [reply GeofencingModule/src/main/java/com/apnamart/geofencingmodule/geofencing/library/GeofenceLibrary.kt]: done

File: GeofencingModule/src/main/java/com/apnamart/geofencingmodule/geofencing/library/GeofenceLibrary.kt (hunk 4 above)
Comment (jinal-apnamart) [reply GeofencingModule/src/main/java/com/apnamart/geofencingmodule/geofencing/library/GeofenceLibrary.kt]: done

File: GeofencingModule/src/main/java/com/apnamart/geofencingmodule/geofencing/permissions/LocationPermissionHelper.kt (hunk 5 above)
Comment (jinal-apnamart) [reply GeofencingModule/src/main/java/com/apnamart/geofencingmodule/geofencing/permissions/LocationPermissionHelper.kt]: done

File: GeofencingModule/src/main/java/com/apnamart/geofencingmodule/geofencing/permissions/LocationPermissionHelper.kt (hunk 6 above)
Comment (jinal-apnamart) [reply GeofencingModule/src/main/java/com/apnamart/geofencingmodule/geofencing/permissions/LocationPermissionHelper.kt]: renamed file

File: GeofencingModule/src/main/java/com/apnamart/geofencingmodule/geofencing/provider/GeofenceDataProvider.kt (hunk 7 above)
Comment (jinal-apnamart) [reply GeofencingModule/src/main/java/com/apnamart/geofencingmodule/geofencing/provider/GeofenceDataProvider.kt]: fixed

File: GeofencingModule/src/main/java/com/apnamart/geofencingmodule/geofencing/work_manager/WorkManagerInitializer.kt (hunk 9 above)
Comment (jinal-apnamart) [reply GeofencingModule/src/main/java/com/apnamart/geofencingmodule/geofencing/work_manager/WorkManagerInitializer.kt:20]: they are for debugging purpose of devs

File: app/build.gradle (hunk 10 above)
Comment (jinal-apnamart) [reply app/build.gradle]: changed it back

File: jitpack.yml (hunk 11 above)
Comment (jinal-apnamart) [reply jitpack.yml]: done

File: pom.xml (hunk 12 above)
Comment (jinal-apnamart) [reply pom.xml]: done

File: settings.gradle (hunk 13 above)
Comment (jinal-apnamart) [reply settings.gradle]: done

**Review by vilas-apnamart — APPROVED** [review]
